connector/jobrunner/runner.py

```diff
@@ -0,0 +1,302 @@
+            self.has_seq = self._has_queue_job_column('seq')
+            self._initialize()
+
+    def close(self):
+        try:
+            self.conn.close()
+        except:
+            pass
+        self.conn = None
+
+    def _has_connector(self):
+        with closing(self.conn.cursor()) as cr:
+            try:
+                cr.execute("SELECT 1 FROM ir_module_module "
+                           "WHERE name=%s AND state=%s",
+                           ('connector', 'installed'))
+            except psycopg2.ProgrammingError as err:
+                if unicode(err).startswith('relation "ir_module_module" '
+                                           'does not exist'):
+                    return False
+                else:
+                    raise
+            return cr.fetchone()
+
+    def _has_queue_job_column(self, column):
+        if not self.has_connector:
+            return False
+        with closing(self.conn.cursor()) as cr:
+            cr.execute("SELECT 1 FROM information_schema.columns "
+                       "WHERE table_name=%s AND column_name=%s",
+                       ('queue_job', column))
+            return cr.fetchone()
+
+    def _initialize(self):
+        with closing(self.conn.cursor()) as cr:
+            # this is the trigger that sends notifications when jobs change
+            # TODO: perhaps we don't need to trigger ON DELETE?
+            cr.execute("""
+                DROP TRIGGER IF EXISTS queue_job_notify ON queue_job;
+
+                CREATE OR REPLACE
+                    FUNCTION queue_job_notify() RETURNS trigger AS $$
+                DECLARE
+                    uuid TEXT;
+                BEGIN
+                    IF TG_OP = 'DELETE' THEN
+                        uuid = OLD.uuid;
+                    ELSE
+                        uuid = NEW.uuid;
+                    END IF;
+                    PERFORM pg_notify('connector',
+                                      current_database() || ',' || uuid);
+                    RETURN NULL;
+                END;
+                $$ LANGUAGE plpgsql;
+
+                CREATE TRIGGER queue_job_notify
+                    AFTER INSERT OR UPDATE OR DELETE
+                    ON queue_job
+                    FOR EACH ROW EXECUTE PROCEDURE queue_job_notify();
+            """)
+            cr.execute("LISTEN connector")
+
+    def select_jobs(self, where, args):
+        query = "SELECT %s, uuid, %s, date_created, priority, eta, state " \
+                "FROM queue_job WHERE %s" % \
+                ('channel' if self.has_channel else 'NULL',
+                 'seq' if self.has_seq else 0,
+                 where)
+        with closing(self.conn.cursor()) as cr:
+            cr.execute(query, args)
+            return list(cr.fetchall())
+
+    def set_job_enqueued(self, uuid):
+        with closing(self.conn.cursor()) as cr:
+            cr.execute("UPDATE queue_job SET state=%s, date_enqueued=NOW() "
```

Comment:
the date_enqueued must be truncated to 'seconds'
```SQL
date_enqueued=date_trunc('seconds', now()::timestamp)
```
otherwise Odoo will not be able to read it and will raise `ValueError`
```
  File "/home/lmi/projects/openerp/openerp-deldrive-buildout/addons-connector/connector/queue/model.py", line 113, in _change_job_state
    job = storage.load(job.uuid)
  File "/home/lmi/projects/openerp/openerp-deldrive-buildout/addons-connector/connector/queue/job.py", line 265, in load
    stored.date_enqueued, DEFAULT_SERVER_DATETIME_FORMAT)
  File "/usr/lib/python2.7/_strptime.py", line 328, in _strptime
    data_string[found.end():])
ValueError: unconverted data remains: .024856
```